### ISSUE TITLE
Mobile ComboBox performance is slow in docs

### ISSUE THREAD

**Issue by LFDanLu** [adobe/react-spectrum#1609]
# 🐛 Bug Report 

There have been reports of sluggish ComboBox performance in the docs. I was able to reproduce it initially but the performance issues went away after restarting Chrome. Performance issues seemed tied to the tray appearance (slow performance for Menu/Picker/etc). Performance in story book was just fine though...

Original tweet thread: https://twitter.com/shcoding/status/1362196684898467841

## 🤔 Expected Behavior

Mobile ComboBox performance shouldn't be slow in docs

## 😯 Current Behavior

Mobile ComboBox performance is slow in docs for some devices (Pixel 2 was mentioned in particular)

## 💁 Possible Solution

Investigate by doing some js profiling in devtools 

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

**Comment by smashercosmo** [comment]
Ok, now I'm pretty sure, that it's not directly related to combobox. I clicked through other component examples and they all behaved very unresponsively. Tabs, Combobox, Picker, Dialog... there are always approximately 2-4 seconds on my phone between clicking a button and component reaction (switching tab, opening dialog or dropdown etc.) So it's either something fundamental in react-spectrum architecture or it's because of the docs site itself.

**Comment by devongovett** [comment]
Possibly related to #721. I assume you're on Android? Perhaps we should make the user-select hack specific to iOS? We'd need to test and make sure text selection behaves as expected on Android, or find a different workaround.

**Comment by smashercosmo** [comment]
Yes, highly likely that this is the reason. I'm on Android (Pixel 2). Also as I said before if I open and close combobox several times Chrome stops responding completely.

**Comment by smashercosmo** [comment]
Just want to say that this is still an issue. Trying to use combobox in the latest docs site on my Pixel 2 still makes Chrome freeze.

**Comment by LFDanLu** [comment]
I'm still unable to reproduce on my device (Galaxy S9+), I assume you are on the latest Chrome version?

**Comment by smashercosmo** [comment]
My chrome version is 96.0.4664.45

**Comment by LFDanLu** [comment]
Hm, same version here. I assume the issue still exists for the other components you mentioned previously (i.e. Tabs, Picker, etc)? Would you mind checking that `user-select: none` is no longer being applied at the page level for you now? It should now be applied on the pressable element since you are on Android.

**Comment by smashercosmo** [comment]
Yes, both Tabs and Picker are affected. Here is the demonstration of this behaviour for Combobox

https://user-images.githubusercontent.com/273283/144907323-2ca9c5b7-7c31-40f0-b644-4aba89cb0f3c.mp4

So, first selection goes well, but the more you interact with component the laggier it behaves

**Comment by smashercosmo** [comment]
Here is how Menu component behaves. It doesn't freeze browser, but there is a noticeable delay between clicking a menu trigger and menu list appearance. And seems like the more you click the longer this delay (but I'm not 100% sure about that)

https://user-images.githubusercontent.com/273283/144908173-3dfa0123-5863-434e-85db-b3231c5d750b.mp4

**Comment by smashercosmo** [comment]
> Would you mind checking that user-select: none is no longer being applied at the page level for you now?

no, I don't see user-select applied neither to html, nor to body elements

**Comment by LFDanLu** [comment]
Thanks for the info, would you mind running the devtools profiler on the ComboBox docs and sharing it here?

**Comment by smashercosmo** [comment]
Do you want me to share screenshot or the actual json?

**Comment by smashercosmo** [comment]
Here is a screenshot
![Screenshot 2021-12-06 at 22 25 12](https://user-images.githubusercontent.com/273283/144917204-f65255fc-f6f0-4a15-8732-63d24d1ad5cc.png)

No quite sure where to put large json file with profiling data

**Comment by smashercosmo** [comment]
is zip fine?
[Profile-20211206T221910.json.zip](https://github.com/adobe/react-spectrum/files/7663554/Profile-20211206T221910.json.zip)

**Comment by LFDanLu** [comment]
That should be fine, I'll reopen this issue so we can investigate

**Comment by LFDanLu** [comment]
One possibility for the slow down could be the `ariaHideOutside` we perform on tray open to hide elements outside of the tray from screen readers, but we have still been unable to reproduce the issue locally (tested with a Pixel 3 recently to no avail). Since this is happening with the ComboBox tray we might be able to go back to the old method where we put `aria-hidden` on the Provider only instead since the combobox input is within the tray and thus will still be visible to screen readers.

@smashercosmo Would you mind helping us test this if we provide you with a copy of the docs with the `ariaHideOutside` behavior altered/removed?

**Comment by smashercosmo** [comment]
Sure, I can test

**Comment by smashercosmo** [comment]
I can also test current version of combobox in storybook, so we're sure that it's either docs related or not docs related issue.

**Comment by LFDanLu** [comment]
@smashercosmo Apologies about the delay, got back from PTO recently. I'm working on getting a working build link for you (verdaccio builds broke after some changes last month), but if you'd like to test the docs locally you can use the changes in this branch to get rid of the `ariaHideOutside` stuff and build the docs via `yarn start:docs --no-cache` and navigating to http://localhost:1234/react-spectrum/ComboBox.html.

**Comment by smashercosmo** [comment]
Sorry, which branch should I use for testing?

**Comment by LFDanLu** [comment]
Forgot to link, sorry about that: https://github.com/adobe/react-spectrum/tree/test_for_issue_1609.

**Comment by LFDanLu** [comment]
Alright, fixed the verdaccio build, you can use https://reactspectrum.blob.core.windows.net/reactspectrum/59cc407b8b1cce78571d51d29d375fa5c9dc7442/verdaccio/docs/react-spectrum/ComboBox.html to test instead of building it locally

**Comment by smashercosmo** [comment]
Yes, on this page Combobox is very performant and doesn't make Chrome freeze

**Comment by LFDanLu** [comment]
Cool, thanks for testing! Mind testing a couple more iterations? Just wanna test the base case and the performance outside of docs.

Base case, should perform the same as prod docs:
https://reactspectrum.blob.core.windows.net/reactspectrum/50ee7418f986e7833688598c18f57c47a041e16e/verdaccio/docs/react-spectrum/ComboBox.html
https://reactspectrum.blob.core.windows.net/reactspectrum/50ee7418f986e7833688598c18f57c47a041e16e/verdaccio/build/index.html

Case where aria-hidden is applied to the Provider:
https://reactspectrum.blob.core.windows.net/reactspectrum/b90ae0fab98066d66247b74370fbfafa2c11214e/verdaccio/docs/react-spectrum/ComboBox.html
https://reactspectrum.blob.core.windows.net/reactspectrum/b90ae0fab98066d66247b74370fbfafa2c11214e/verdaccio/build/index.html

**Comment by smashercosmo** [comment]
Both cases make chrome freeze

**Comment by smashercosmo** [comment]
Outside of docs works fine in both cases

**Comment by LFDanLu** [comment]
Awesome, thanks so much for verifying those cases. I'll have to dig some more to see if there is an alternative to the `aria-hidden` on external Provider idea then

**Comment by izznatsir** [comment]
Considering that it does not occur on browser other that chromium-based. Might be better to file an issue to chromium about this. I also experience the sam issue on my android phone.

**Comment by LFDanLu** [comment]
I didn't want to file an issue against chromium because the issue was only reproducible in our docs, still need to dig into why.

**Comment by badtant** [comment]
I'm also experiencing this on Galaxy S10e. Happy to help with testing if needed.
Thanks

**Comment by jussivirtanen** [comment]
I've also noticed this `ComboBox` slowness on desktop Chrome, although less than in the examples others have shown in this issue.

Here's a screen recording showing Chrome (on the left) being laggy with the Combobox. Firefox (on the right) and Safari work as you'd imagine the component should work.

https://user-images.githubusercontent.com/3430121/193762775-ae5a8523-d795-4f93-b1cc-c5dd0658f235.mp4

**Details**
Browser: Chrome Version 106.0.5249.91 (Official Build) (arm64)
OS: macOS Monterey Version 12.6

**Comment by LFDanLu** [comment]
@jussivirtanen Thanks for the video, are you able to reproduce in this sandbox: https://codesandbox.io/s/hardcore-moon-xzc4r?file=/src/ComboBox.tsx? I haven't been able to reproduce what you are experiencing yet in the docs or the sandbox but definitely something I can keep an eye on when I get some time to dig into this issue some more.

**Comment by omridevk** [comment]
@LFDanLu  isn't using this https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert instead of aria hidden could solve the issue?

**Comment by LFDanLu** [comment]
That's a good idea, something we've tried via https://github.com/adobe/react-spectrum/pull/4314 but resulted in [several](https://github.com/adobe/react-spectrum/pull/4942) [bugs](https://github.com/adobe/react-spectrum/pull/4881) which we then needed to revert the PR changes. Unfortunately `inert` causes elements to become completely non-clickable/focusable so it isn't a 1:1 replacement for our `aria-hidden` strategy where we want to hide elements outside of modals/dropdowns from screen readers but keep them focusable in case the user wants to tab out of them.

**Comment by omridevk** [comment]
not sure I follow, I thought that modals/combobox which have focus trap shouldn't allow focusing outside of the region. For dropdown not sure why would you want to use focus trap.

**Comment by LFDanLu** [comment]
Our approach to hiding all elements outside via `ariaHideOutside` from screen readers isn't strictly tied to whether or not the popover/modal/dropdown is a focus trap, it is also in part to ensure screenreader users can move smoothly from the trigger element to the portalled popover/modal/dropdown, see [here](https://github.com/adobe/react-spectrum/discussions/4871#discussioncomment-6673489) for some more details.

**Comment by omridevk** [comment]
Love it! awesome work @devongovett
